design/ErrorHandling.md

```diff
@@ -23,6 +23,141 @@ There are several types of WebGPU calls that get their errors handled differentl
 Implementations should provide a way to enable synchronous validation, for example via a debug shim or via the developer tools.
 The extra overhead needs to be low enough that applications can still run while being debugged.
 
+## *Fatal Errors*: Lost/Recovered Events
+
+<!-- calling this revision 5 -->
+
+```webidl
+partial interface GPUAdapter {
+    Promise<GPUDevice> requestDevice(GPUDeviceDescriptor descriptor,
+                                     GPUDeviceLostCallback onDeviceLost);
+};
+
+callback GPUDeviceLostCallback = void (GPUDeviceLostInfo info);
+
+interface GPUDeviceLostInfo {
+    readonly attribute GPUDevice device;
+    readonly attribute DOMString reason;
+};
+```
+
+`GPUAdapter.requestDevice` requests a device from the adapter.
+It returns a Promise which resolves when a device is ready.
+The Promise may not resolve for a long time - it resolves when the browser is ready for the application to bring up (or restore) its content.
+If the adapter is lost before the Promise resolves, the Promise rejects.
+
+`requestDevice` takes a required `onDeviceLost` callback.
+When it is called, the `GPUDevice` cannot be used anymore.
+The device and all objects created from the device have become invalid.
+All further operations on the device and its objects are errors.
+The `"gpulogentry"` event will no longer fire. (This makes all further operations no-ops.)
+(TODO: this could probably be an `ondevicelost` event instead.
+However, this (1) makes it required and (2) makes it clear that no error can be missed between creation and event listener registration.)
+
+`onDeviceLost` may be called if something goes fatally wrong on the device (e.g. unexpected out-of-memory, crash, or native device loss).
+When the application receives this callback, it may immediately request a new device.
+
+### Example Code
+
+```js
+class MyRenderer {
+  constructor() {
+    this.rendering = false;
+    this.adapter = null;
+    this.device = null;
+  }
+  async begin() {
+    try {
+      await this.initWebGPU();
+    } catch (e) {
+      console.error(e);
+      this.initFallback();
+    }
+  }
+  async initWebGPU() {
+    await this.ensureDevice();
+    // ... Upload resources, etc.
+  }
+  initFallback() { /* ... */ }
+  async ensureDevice() {
+    // Stop rendering. (If there was already a device, WebGPU calls made before
+    // the app notices the device is lost are okay - they are no-ops.)
+    this.rendering = false;
+    if (!this.adapter) {
+      // If no adapter, get one.
+      // (If requestAdapter rejects, no matching adapter is available. Exit to fallback.)
+      this.adapter = await gpu.requestAdapter({ /* options */ });
+    }
+    try {
+      this.device = await this.adapter.requestDevice({ /* options */ }, (info) => {
+        // Device was lost.
+        console.error("device lost", info);
+        // Try to get a device again.
+        this.ensureDevice();
```

Comment:
This said, the sample code *could* just always try on a newly requested adapter instead of retrying on the same adapter before trying on a new adapter. Both would be valid choices for an app, but the distinction also seems minimal. Maybe it would be better for the sample to be simpler.

Reply:
Decided not to simplify the example, because it's useful to demonstrate the use case. However, I did refactor the example to be iterative instead of recursive, which is clearer. I think it is equivalent.